src/main/java/frc/robot/subsystems/LogSubsystem.java

```diff
@@ -0,0 +1,153 @@
+package frc.robot.subsystems; 
+ 
+import java.io.File;
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+ 
+import edu.wpi.first.wpilibj.DriverStation;
+ 
+public class LogSubsystem {
+    public String m_subsystemName;
+    private static LogSubsystem INSTANCE = new LogSubsystem();
+    public String m_level;
+    public static final String RED = "\033[0;31m";     // RED
+    public static final String BLUE = "\033[0;34m";    // BLUE
+    
+    public static LogSubsystem getInstance() {
+        return INSTANCE;
+    }
+ 
+ 
+    public enum Level {
+        kRobot, kInfo, kWarning, kLibrary
+    }
+    private final List<LogSource> dataSources = new ArrayList<>();
+    private Path file;
+ 
+    private String loggingLocation = "/home/lvuser/logs/";
+ 
+    private LogSubsystem() {
+        File usb1 = new File("/media/sda1/");
+        if (usb1.exists()) {
+            loggingLocation = "/media/sda1/logs/";
+        }
+    }
+ 
+    private void createLogDirectory() throws IOException {
+        File logDirectory = new File(loggingLocation);
+        if (!logDirectory.exists()) {
+            Files.createDirectory(Paths.get(loggingLocation));
+        }
+    }
+ 
+    private void createFile() {
+        Writer output = null;
+        try {
+            createLogDirectory();
+            if (DriverStation.getInstance().isFMSAttached()) {
+                file = Paths.get(loggingLocation + "_" + m_subsystemName + DriverStation.getInstance().getEventName()
+                        + "_" + DriverStation.getInstance().getMatchType()
+                        + DriverStation.getInstance().getMatchNumber() + ".csv");
+            } else {
+                file = Paths.get(loggingLocation + "test.csv");
+            }
+            if (Files.exists(file)) {
+                Files.delete(file);
+            }
+            Files.createFile(file);
+            saveTitles();
+        } catch (Exception e) {
+            e.printStackTrace();
+        } finally {
+            if (output != null) {
+                try {
+                    output.close();
+                } catch (IOException e) {
+                }
+            }
+        }
+    }
+ 
+    public void addSource(String name, String subsystemName, Supplier<Object> supplier) {
+    
+        dataSources.add(new LogSource(name,  subsystemName,supplier));
+    }
+ 
+    public void saveLogs() {
+        try {
+            if (file == null) {
+                createFile();
+            }
+ 
+            StringBuilder data = new StringBuilder();
+            data.append(m_level);
```

Comment:
şurada m_level şeklinde almak doğru çalışmayabilir, çünkü her source'un levelı farklı olacak. bu şekilde sadece son belirlenen levelı gösterir her source için. getValues fonksiyonunun getLevels halini yapıp onu kullanman gerekiyor.

Reply:
diğer branch'de bulunan level kısmına baktın mı